### PR TITLE
Add borg extra arguments (#96)

### PR DESCRIPTION
**Pull request by ethernet-zero**
Added a field to the repo information so additional arguments can be
added to the borg command line. The arguments specified there are
appended to every command that's run on that repo.

### CODE REVIEW

File: src/vorta/views/repo_tab.py [hunk 1]
```diff
@@ -129,7 +129,7 @@ def ssh_copy_to_clipboard_action(self):
 
     def compression_select_action(self, index):
         profile = self.profile()
-        profile.compression = self.repoCompression.currentData()
+        profile.compression = self.repoCompression.currentData() or ''
```
Comment (ThomasWaldmann) [inline src/vorta/views/repo_tab.py]: this seems to be unrelated to the topic of your PR.

Comment (ethernet-zero) [reply src/vorta/views/repo_tab.py]: Oops, you're right. I had to add that because after applying migrations to the settings database the program ended with an error on that line because of the NOT NULL constraint on that field.

Should I remove that file from the PR?

Comment (m3nu) [reply src/vorta/views/repo_tab.py]: We recently changed the available compression options. They should be migrated when you run a version later than commit 3efecd71f3b0fc69261284efcd8d628bb2b89d59.

File: src/vorta/borg/borg_thread.py [hunk 2]
```diff
@@ -142,7 +144,8 @@ def run(self):
         log_entry.save()
         logger.info('Running command %s', ' '.join(self.cmd))
 
-        p = Popen(self.cmd, stdout=PIPE, stderr=PIPE, bufsize=1, universal_newlines=True,
+        extra_args = shlex.split(self.params.get('extra_borg_arguments', ''))
+        p = Popen(self.cmd + extra_args, stdout=PIPE, stderr=PIPE, bufsize=1, universal_newlines=True,
```
Comment (m3nu) [inline src/vorta/borg/borg_thread.py]: This appends extra args at the very end. Like `borg create --debug $REPO::$SNAPSHO --rsh=xxx`.

Another option would be to append them in `vorta.borg.borg_thread.prepare_bin()`. Because extra args apply to *all* different `borg` commands, right?

Comment (ethernet-zero) [reply src/vorta/borg/borg_thread.py]: They do.

Sure, it'd work either way. I added them at the end to not change the current interface of `prepare_bin()`.

Comment (ThomasWaldmann) [reply src/vorta/borg/borg_thread.py]: please check "order matters" in the borg docs.

Comment (m3nu) [reply src/vorta/borg/borg_thread.py]: Why not insert the extra args around [Line 44 in this file](https://github.com/borgbase/vorta/blob/master/src/vorta/borg/borg_thread.py#L44). This is where global options are added. Python also has a helper function to insert elements into a list, like `self.cmd`

```
# (existing) change path to Borg binary, if needed
cmd[0] = self.prepare_bin()

# insert extra args
cmd.insert(1, params.['extra_borg_arguments'])  # TODO: skip if no extra args, split args
```

Comment (ethernet-zero) [reply src/vorta/borg/borg_thread.py]: I agree. @ThomasWaldmann you're right about the order of the arguments, I fixed it.

@m3nu I moved constructing the command line to where you suggested. 👍

**Comment by m3nu** [comment]
We're getting there. When testing I noticed:

- The "Add/Edit Repo" repo window/sheet is fairly overloaded with the new field. Most users don't need this field at all and may be confused by it. So I suggest to hide it under "Advanced" or something. Need to see what options Qt offers here. I'm sure there area a few.
- The field label says **Extra Borg arguments:**, but CONTRIBUTING.md says: *Field labels (same or next line) end with a colon and are titleized. "Allowed Networks:"*. So should be **Extra Borg Arguments:**

**Comment by m3nu** [comment]
Tests are passing now. I wonder if there should be a test to see if the Borg command was correctly contructed given a repo with some custom commands.

**Comment by ethernet-zero** [comment]
>     * The "Add/Edit Repo" repo window/sheet is fairly overloaded with the new field. Most users don't need this field at all and may be confused by it. So I suggest to hide it under "Advanced" or something together with the SSH dropdown. Need to see what options Qt offers here. I'm sure there area a few.
If we moved the field to the Advanced tab, its contents would have to change with the selected repo, right? It'd make sense to put it there along with the pre- and post-backup commands, but since it'd be tied to the repo and its contents would change, it could be a good idea to clarify that in a small help text label or something like that.

>     * The field label says **Extra Borg arguments:**, but CONTRIBUTING.md says: _Field labels (same or next line) end with a colon and are titleized. "Allowed Networks:"_. So should be **Extra Borg Arguments:**
Right, fixing it.

**Comment by m3nu** [comment]
> If we moved the field to the Advanced tab, its contents would have to change with the selected repo, right? It'd make sense to put it there along with the pre- and post-backup commands, but since it'd be tied to the repo and its contents would change, it could be a good idea to clarify that in a small help text label or something like that.

I didn't mean moving it to some tab. I meant adding a small label "Advanced" in the **same** window and only showing the SSH key- and extra args fields when the user clicks it.

**Comment by m3nu** [comment]
screen shot 2019-02-19 at 18 47 07

A *Basic*/*Advanced* tab could work. (I'm not saying that you need to code everything yourself. We just need a good solution. Implementation is usually the easy part)

**Comment by ethernet-zero** [comment]
Ah, alright, got it. :smile:

**Comment by m3nu** [comment]
Maybe the "Encryption" selection is advanced too. So only URL and Password are required. That could work and we can add more advanced options later.

Coding-wise the layout changes don't affect the code at all, as long as the field names stay the same.

**Comment by ethernet-zero** [comment]
Could this work?




I left the SSH key selection in the Basic tab to not leave one tab with three fields and another tab with just one, but it could be moved to Advanced just as easily.

**Comment by ethernet-zero** [comment]
And this would be when moving the SSH key field to the Advanced tab:




Which one do you prefer?

**Comment by m3nu** [comment]
Having the URL separate on top seems unusual and makes it harder to understand. I'd think about it in terms of Required + optional parameters. When moving the URL up, you are creating a third group.

How about:

- Tab *General*/*Required*/*Basic*/...: URL, password
- Tab *Advanced*: SSH, extra params, encryption

I looked into some other apps I use to see how they solved it. Also to point out that this is not about *my* preference. It needs to make sense for whoever is using it to enable them to quickly do the task they want to do.

screen shot 2019-02-20 at 13 07 57
screen shot 2019-02-20 at 13 09 26
screen shot 2019-02-20 at 13 11 43
screen shot 2019-02-20 at 13 12 32

**Comment by ethernet-zero** [comment]
There you go:

**Comment by ethernet-zero** [comment]
I fixed the lines that build the borg command line so the extra arguments are placed right after the subcommand, because putting them after the borg binary messed up the activity log.

**Comment by m3nu** [comment]
Rebased and tried to clean up the repo-add window a bit.

**Comment by m3nu** [comment]
While working on another issue, I had Vorta hanging before every call to Borg. Seems to happen when `shlex.split` is called on `None`. Same thing happens in iPython.

```
In [1]: import shlex
In [2]: shlex.split(None)  # will hang here
```

Since the DB field was set to allow null, `None` was getting passed. I changed this now to only have the default value of an empty string.

I didn't see this hanger before, but I'll include this change in another PR just in case.

**Comment by ethernet-zero** [comment]
It seems that passing `None` to `shlex.split()` makes it read from the standard input, that's why it hangs. Good catch. :smile:
